### PR TITLE
nGraph Transformations refactoring

### PR DESCRIPTION
**Pull request by GlebKazantaev**
This PR introduces next changes:
1. Transformations *_tbl.hpp files were replaced with direct registration in cpp files.
2. Plugins use pass::Manager to call conversion passes.
3. Transformations callback was moved to PassBase class as there is no more need to keep it in separate class
4. All pattern based transformations must be inherited from MatcherPass class. GraphRewrite class will be used only for matchers registration and execution on function.
MatcherPass class adds new features to pattern-based transformations approach:
* Allows to run matcher pass on a single node.
* Operations that were created inside transformation callback can be added to execution list to be available for pattern matching within single GraphRewrite.
5. GraphRewrite MatchClosure was replaced with MatcherPass. So all matchers will be registered as a MatcherPass.
6. Added pass::Manager::clear_state() method to avoid dependency with nodes that no longer belongs to function after replacement.
7.  Some representative transformations were updated to use MatcherPass as an example.
8.  Mul->Add sequence fusion transformation was replaced with LinOpSequenceFusion.
9. Pattern and callback registration code was moved to class c-tors (will be finished for remaining passes in other PR) .
10. Updated pass::Manager to get pass names only when NGRAPH_PROFILE_PASS_ENABLE enabled.
11. Moving towards removing PassProperty.
12. Added ngraph::pattern::wrap_type<T>(inputs, pred) to simplify pattern creation.
13. GraphRewrite was updated to execute MatcherPass more efficient.

### CODE REVIEW

**Comment by GlebKazantaev** [comment]
For random model it gives conversion speedup from 1.5sec to 1.1sec

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp [hunk 1]
```diff
@@ -50,22 +50,73 @@
...
+
+
+    auto anchor = manager.register_pass<ngraph::pass::GraphRewrite>();
+    anchor->add_matcher<ngraph::pass::ConvertNormalizeL2ToLegacyMatcher>();
+    anchor->add_matcher<ngraph::pass::ConvertHardSigmoidToLegacyMatcher>();
+    anchor->add_matcher<ngraph::pass::ConvertProposalToLegacyMatcher>();
+    anchor->add_matcher<ngraph::pass::ConvertTileToLegacyMatcher>();
+    anchor->add_matcher<ngraph::pass::ConvertLRNToLegacyMatcher>();
+    anchor->add_matcher<ngraph::pass::ConvertPadToLegacyMatcher>();
+    anchor->set_name("ngraph::pass::ConvertOpSet1ToLegacy");
```
Comment (slyalin) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:101]: Why "ConvertOpSet1ToLegacy"? It is only a part of it.

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp [hunk 2]
```diff
@@ -50,22 +50,73 @@
...
+    manager.register_pass<ngraph::pass::ConstantFolding>();
+    manager.register_pass<ngraph::pass::ConvertReduceToPooling>();
+    manager.register_pass<ngraph::pass::ConvertMod>();
+    manager.register_pass<ngraph::pass::ConvertMinimum>();
+    manager.register_pass<ngraph::pass::ConvertSubtract>();
+    manager.register_pass<ngraph::pass::ConvertDivide>();
+    manager.register_pass<ngraph::pass::ConvertNegative>();
+    manager.register_pass<ngraph::pass::ConvertDepthToSpace>();
+    manager.register_pass<ngraph::pass::ConvertSpaceToDepth>();
+    manager.register_pass<ngraph::pass::ConvertConvolutions>();
```
Comment (slyalin) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:71]: All these passes look capable to be run in a single pass through the function. You just didn't have time to reorganize or it is blocked by something?

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp (hunk 1 above)
Comment (slyalin) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:101]: Is it only example of "merged" passes that we are going to have in ConvertToLegacy or is it only the first example and it will be further reorganized?

Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:101]: It's just usage example. Ideally all transformations below also can be merged into single GraphRewrite pass

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp (hunk 2 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:71]: It requires deeper investigation because some of this transformations will be moved to CommonOptimizations, some of them have dependencies on execution order and some of them can be merged into single pass.

File: ngraph/src/ngraph/pass/graph_rewrite.hpp [hunk 3]
```diff
@@ -82,6 +110,33 @@ class NGRAPH_API ngraph::pass::GraphRewriteBase : public ngraph::pass::FunctionP
 class NGRAPH_API ngraph::pass::GraphRewrite : public ngraph::pass::GraphRewriteBase
 {
 public:
+    GraphRewrite()
+        : GraphRewriteBase()
```
Comment (slyalin) [inline ngraph/src/ngraph/pass/graph_rewrite.hpp:114]: Again calling base class default c-tor for... trivial c-tor. Recommend to replace by `= default`.

File: ngraph/src/ngraph/pass/graph_rewrite.cpp [hunk 4]
```diff
@@ -66,49 +66,51 @@ bool pass::GraphRewrite::run_on_function(shared_ptr<Function> f)
...
                 {
                     rewritten = true;
-                    // If call back may change function's is_dynamic state, we need to
-                    // update the cached value.
-                    if (closure.property.is_set(PassProperty::CHANGE_DYNAMIC_STATE))
+                    if (m_pass->get_matcher_property().is_set(
+                            PassProperty::REQUIRE_SHAPE_INFERENCE))
                     {
-                        is_dyn_func = s_rerun_dynamic_check && f->is_dynamic();
+                        f->validate_nodes_and_infer_types();
```
Comment (slyalin) [inline ngraph/src/ngraph/pass/graph_rewrite.cpp:107]: Is it too costly to evaluate entire function? Can we provide a way for MatcherPass to report about place in a function where the earliest (according to data-flow) changes happened to initiate shape inference at some point in the graph. Also, `Function::validate_and_infer_types` has too straight-forward implementation for our need here: it doesn't stop when shape and type inferred for the second time match completely the original ones. Can we think (probably in the following PR) about a way to avoid entire function re-evaluation?

File: ngraph/src/ngraph/pass/manager.cpp [hunk 5]
```diff
@@ -63,69 +60,70 @@ void pass::Manager::run_passes(shared_ptr<Function> func, bool /* transitive */)
...
             {
                 vt_pass->set_ops_to_details(get_state().get_visualize_tree_ops_map());
             }
             module_pass->run_on_module(f_array);
         }
-        else if (function_pass)
+        else if (auto matcher_pass = dynamic_pointer_cast<MatcherPass>(pass))
         {
-            for (auto f_pair : fs)
+            // This checks is to skip the graph optimization when the graph pass relies on
```
Comment (slyalin) [inline ngraph/src/ngraph/pass/manager.cpp:78]: `optimization` --> `transformation`

File: ngraph/src/ngraph/pass/manager.hpp [hunk 6]
```diff
@@ -63,6 +63,12 @@ class NGRAPH_API ngraph::pass::Manager
     /// each registered pass
     /// \param new_state Value "true" enables Validate pass run; "false", otherwise
     void set_per_pass_validation(bool new_state) { m_per_pass_validation = new_state; }
+    void set_callback(param_callback callback)
```
Comment (slyalin) [inline ngraph/src/ngraph/pass/manager.hpp:66]: Please explain the purpose of this function in comments.

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_pad_to_pad_ie.cpp [hunk 7]
```diff
@@ -10,14 +10,8 @@
...
-void ngraph::pass::ConvertPadToPadIE::convert_pad() {
-    auto input_0 = std::make_shared<pattern::op::Label>(element::f32, Shape{1, 1, 1, 1});
-    auto input_1 = std::make_shared<pattern::op::Label>(element::i64, Shape{4});
-    auto input_2 = std::make_shared<pattern::op::Label>(element::i64, Shape{4});
-    auto input_3 = std::make_shared<pattern::op::Label>(element::f32, Shape{});
-    auto pad_1 = std::make_shared<ngraph::opset1::Pad>(input_0, input_1, input_2, op::PadMode::SYMMETRIC);
-    auto pad_2 = std::make_shared<ngraph::opset1::Pad>(input_0, input_1, input_2, input_3, op::PadMode::CONSTANT);
-
+void ngraph::pass::ConvertPadToLegacyMatcher::convert_pad() {
+    auto pad = std::make_shared<pattern::op::Label>(element::f32, Shape{}, pattern::has_class<ngraph::opset1::Pad>());
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_pad_to_pad_ie.cpp:14]: It's a different story and I planned it after this changes.

File: inference-engine/src/transformations/include/transformations/convert_depth_to_space.hpp [hunk 8]
```diff
@@ -20,9 +19,9 @@ class TRANSFORMATIONS_API ConvertDepthToSpace;
 }  // namespace pass
 }  // namespace ngraph
 
-class ngraph::pass::ConvertDepthToSpace: public ngraph::pass::GraphRewrite, public ngraph::pass::PassParam {
+class ngraph::pass::ConvertDepthToSpace: public ngraph::pass::GraphRewrite {
 public:
-    ConvertDepthToSpace() : GraphRewrite(), PassParam() {
+    ConvertDepthToSpace() : GraphRewrite() {
         convert_depth_to_space();
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply inference-engine/src/transformations/include/transformations/convert_depth_to_space.hpp:25]: Agree

File: inference-engine/src/transformations/include/transformations/convert_opset1_to_legacy/convert_tile_to_ie_tile.hpp [hunk 9]
```diff
@@ -14,14 +14,14 @@
...
+class TRANSFORMATIONS_API ConvertTileToLegacyMatcher;
 
 }  // namespace pass
 }  // namespace ngraph
 
-class ngraph::pass::ConvertTileToIETile: public ngraph::pass::GraphRewrite {
+class ngraph::pass::ConvertTileToLegacyMatcher: public ngraph::pass::MatcherPass {
 public:
-    ConvertTileToIETile() : GraphRewrite() {
+    ConvertTileToLegacyMatcher() : MatcherPass() {
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply inference-engine/src/transformations/include/transformations/convert_opset1_to_legacy/convert_tile_to_ie_tile.hpp:24]: Agree

File: inference-engine/src/transformations/include/transformations/convert_opset1_to_legacy/convert_hard_sigmoid_to_hard_sigmoid_ie.hpp [hunk 10]
```diff
@@ -15,14 +15,14 @@
...
 
 }  // namespace pass
 }  // namespace ngraph
 
-class ngraph::pass::ConvertHardSigmoidToHardSigmoidIE : public ngraph::pass::GraphRewrite {
+class ngraph::pass::ConvertHardSigmoidToLegacyMatcher : public ngraph::pass::MatcherPass {
 public:
-    ConvertHardSigmoidToHardSigmoidIE() : GraphRewrite() {
+    ConvertHardSigmoidToLegacyMatcher() : MatcherPass() {
         convert_hard_sigmoid();
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply inference-engine/src/transformations/include/transformations/convert_opset1_to_legacy/convert_hard_sigmoid_to_hard_sigmoid_ie.hpp:26]: Agree

File: inference-engine/src/transformations/include/transformations/convert_depth_to_space.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply inference-engine/src/transformations/include/transformations/convert_depth_to_space.hpp:25]: I'll move matchers not to use this function. Other passes will be updated later in other PR.

File: ngraph/src/ngraph/pass/graph_rewrite.hpp [hunk 11]
```diff
@@ -30,13 +30,47 @@ namespace ngraph
...
+
+    struct MatchClosure
+    {
+        std::string name;
+        std::function<bool(const std::shared_ptr<Node>& node)> handler;
+        pass::PassPropertyMask property;
+    };
 }
 
+class NGRAPH_API ngraph::pass::MatcherPass : public ngraph::pass::PassBase
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/graph_rewrite.hpp:48]: I'll add Doxygen comments here

File: ngraph/src/ngraph/pass/graph_rewrite.hpp (hunk 3 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/graph_rewrite.hpp:114]: done.

File: ngraph/src/ngraph/pass/graph_rewrite.hpp [hunk 12]
```diff
@@ -30,13 +30,47 @@ namespace ngraph
...
+    explicit MatcherPass(const MatchClosure& match)
+        : PassBase()
+    {
+        m_matcher = match;
+    }
+
+    bool apply_matcher(std::shared_ptr<ngraph::Node> node);
+
+    const std::string& get_matcher_name() const { return m_matcher.name; }
+    const PassPropertyMask get_matcher_property() const { return m_matcher.property; }
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/graph_rewrite.hpp:65]: Changed.

File: ngraph/src/ngraph/pass/graph_rewrite.cpp [hunk 13]
```diff
@@ -260,3 +263,25 @@ bool pass::RecurrentGraphRewrite::run_on_function(shared_ptr<Function> f)
...
+                     NGRAPH_DEBUG << "Running matcher " << m->get_name() << " on " << node;
+                     if (m->match(node->output(0)))
+                     {
+                         NGRAPH_DEBUG << "Matcher " << m->get_name() << " matched " << node;
+                         return callback(*m.get());
+                     }
+                     return false;
+                 },
+                 property};
+}
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/graph_rewrite.cpp:282]: Changed to register_matcher

File: ngraph/src/ngraph/pass/graph_rewrite.hpp [hunk 14]
```diff
@@ -30,13 +30,47 @@ namespace ngraph
...
 
+class NGRAPH_API ngraph::pass::MatcherPass : public ngraph::pass::PassBase
+{
+public:
+    MatcherPass()
+        : PassBase()
+    {
+    }
+
+    explicit MatcherPass(const MatchClosure& match)
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/graph_rewrite.hpp:56]: Agree. Changed.

File: ngraph/src/ngraph/pass/manager.cpp (hunk 5 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/manager.cpp:78]: done.

File: ngraph/src/ngraph/pass/graph_rewrite.cpp (hunk 4 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/graph_rewrite.cpp]: This was an experimental feature, we decided not to include it as is't not needed. Also PassProperty is going to be removed. Partial re-evaluatoin is a good feature, but lets first understand use cases and potential ways of implementation. I would keep  it as a separate feature/PR.

File: ngraph/src/ngraph/pass/manager.hpp (hunk 6 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/manager.hpp:83]: done.

File: docs/IE_PLUGIN_DG/NewTransformation.md [hunk 15]
```diff
@@ -139,10 +177,12 @@ This example show how we can construct pattern when operation has arbitrary numb
...
 @snippet example_ngraph_utils.cpp pattern:concat_example
 
-This example shows how to use predicate to construct pattern where operation has two different types.
+This example shows how to use predicate to construct pattern where operation has two different types. Also it shows how to match pattern manually on given node.
 
 @snippet example_ngraph_utils.cpp pattern:predicate_example
 
+> **Note**: be careful with manual matching because Matcher object holds matched nodes. To clear match use m->clear_state() method.
+
 TODO: add examples for ngraph::pattern::op::Any
```
Comment (ilya-lavrenov) [inline docs/IE_PLUGIN_DG/NewTransformation.md:186]: should this `TODO` be fixed in this PR?

File: docs/IE_PLUGIN_DG/NewTransformation.md [hunk 16]
```diff
@@ -353,26 +392,34 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
-
-// Update callback to be able to use transformation_callback if this transformation based on GraphRewrite.
+// Update callback to be able to use m_transformation_callback if this transformation based on GraphRewrite.
 ngraph::graph_rewrite_callback callback = [this](pattern::Matcher &m) {
     ...
 }
 
-// Use transformation_callback not to execute transformation
-if (transformation_callback(node)) {
+// Use transformation_callback not to execute transformation if callback returns true fro given node
```
Comment (ilya-lavrenov) [inline docs/IE_PLUGIN_DG/NewTransformation.md:403]: `fro` -> `for`

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp [hunk 17]
```diff
@@ -46,26 +45,103 @@
...
+    fusion->set_name("ngraph::pass::Fusions");
+
+    // CF is required after fusions
+    manager.register_pass<ngraph::pass::ConstantFolding>();
+
+    // List of passes that convert opset1 operations to legacy
+    // plus transformations that are required by InferenceEngine
+    // All this transformations can be executed simultaneously
+    auto anchor = manager.register_pass<ngraph::pass::GraphRewrite>();
+    anchor->add_matcher<ngraph::pass::ReshapeFullyConnected>();
```
Comment (ilya-lavrenov) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:108]: can `manager` merge 2 or more `GraphRewrite` passes? and use single pass through the graph instead of 2

File: docs/IE_PLUGIN_DG/NewTransformation.md [hunk 18]
```diff
@@ -353,26 +392,34 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
+const auto transformations_callback = [](const std::shared_ptr<const ::ngraph::Node> &node) -> bool {
+    return std::dynamic_pointer_cast<const ::ngraph::opset3::StridedSlice>(node) != nullptr;
+};
+
+// Register transformations and pass callback to pass::Manager
+ngraph::pass::Manager manager;
+manager.register_pass<ngraph::pass::CommonOptimizations>();
+...
+// ConvertOpSet1ToLegacy contains ConvertStridedSliceToCrop transformation
+manager.register_pass<ngraph::pass::ConvertOpSet1ToLegacy>();
```
Comment (ilya-lavrenov) [inline docs/IE_PLUGIN_DG/NewTransformation.md:418]: Do we need to demonstrate usage of legacy passes in docs? I suppose - no

File: docs/template_plugin/src/template_pattern_transformation.hpp [hunk 19]
```diff
@@ -12,20 +12,21 @@
...
+class DecomposeDivideMatcher;
+class ReluReluFusionMatcher;
 
 }  // namespace pass
 }  // namespace ngraph
 
 // ! [graph_rewrite:template_transformation_hpp]
 // template_pattern_transformation.hpp
-class ngraph::pass::MyPatternBasedTransformation: public ngraph::pass::GraphRewrite {
+class ngraph::pass::DecomposeDivideMatcher: public ngraph::pass::MatcherPass {
```
Comment (ilya-lavrenov) [inline docs/template_plugin/src/template_pattern_transformation.hpp:23]: `MatcherPass` -> `PatternMatcherPass`?

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_strided_slice_to_crop.cpp [hunk 20]
```diff
@@ -22,7 +22,7 @@ void ngraph::pass::ConvertStridedSliceToCrop::convert_strided_slice_to_crop() {
     std::vector<int64_t> end_mask = {0, 0, 0, 0};
     auto m_slice = std::make_shared<ngraph::opset1::StridedSlice>(data, m_begin, m_end, m_stride, begin_mask, end_mask);
 
-    ngraph::graph_rewrite_callback callback = [](pattern::Matcher& m) {
+    ngraph::matcher_pass_callback callback = [](pattern::Matcher& m) {
```
Comment (ilya-lavrenov) [inline inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_strided_slice_to_crop.cpp:25]: `matcher_pass_callback` -> `pattern_matcher_callback`?

File: ngraph/src/ngraph/pass/pass.hpp [hunk 21]
```diff
@@ -32,8 +38,8 @@ namespace ngraph
         class PassBase;
         class ModulePass;
         class FunctionPass;
-        class NodePass;
-        class CallGraphPass;
+        class NodePass NGRAPH_DEPRECATED("Use MatcherPass or FunctionPass instead.");
+        class CallGraphPass NGRAPH_DEPRECATED("Use MatcherPass or FunctionPass instead.");
```
Comment (ilya-lavrenov) [inline ngraph/src/ngraph/pass/pass.hpp:42]: @ilyachur @GlebKazantaev when it's going to be removed? only in 2022.1? Or we can remove it's usage and remove even in this release?

File: ngraph/src/ngraph/pass/pass.hpp [hunk 22]
```diff
@@ -79,14 +87,45 @@ class NGRAPH_API ngraph::pass::PassBase
...
+    std::string get_name() const
+    {
+        if (m_name.empty())
+        {
+            const PassBase* p = this;
+            std::string pass_name = typeid(*p).name();
+#ifndef _WIN32
+            int status;
+            pass_name = abi::__cxa_demangle(pass_name.c_str(), nullptr, nullptr, &status);
+#endif
```
Comment (ilya-lavrenov) [inline ngraph/src/ngraph/pass/pass.hpp:100]: is it portable? Looks strange

File: ngraph/src/ngraph/pass/manager.hpp [hunk 23]
```diff
@@ -63,6 +63,29 @@ class NGRAPH_API ngraph::pass::Manager
...
+    /// This callback returns true in case of DepthToSpace operation. So when execution DepthToSpace
+    /// decomposition pass will check is this decomposition needed or plugin can execute this
+    /// operation directly. And of course on transformation side we need to have a response for this
+    /// callback.
+    /// if (m_transformation_callback(batch_to_space)) {
+    ///     return false;
+    /// }
+    /// \param callback lamda function that returns true in case if node is supported by plugin and
+    /// transformation is not needed
+    void set_callback(param_callback callback)
```
Comment (ilya-lavrenov) [inline ngraph/src/ngraph/pass/manager.hpp:83]: Should not it be named as `ParamCallback` according to ngraph naming convention

File: ngraph/src/ngraph/pass/pass.hpp (hunk 22 above)
Comment (ilya-lavrenov) [reply ngraph/src/ngraph/pass/pass.hpp:100]: at least before this change, it was hidden inside CPP..

File: inference-engine/tests/unit/inference_engine/ngraph_graph_rewrite_test.cpp [hunk 24]
```diff
@@ -0,0 +1,97 @@
+// Copyright (C) 2018-2020 Intel Corporation
```
Comment (ilya-lavrenov) [inline inference-engine/tests/unit/inference_engine/ngraph_graph_rewrite_test.cpp:1]: why do we have these tests inside IE, not inside ngraph?

File: docs/IE_PLUGIN_DG/NewTransformation.md [hunk 25]
```diff
@@ -2,21 +2,28 @@
...
+Transformations library is independent from InferenceEngine target library named as `inference_engine_transformations`
+and located in `inference-engine/src/transformations` directory.
+
 Transformations root directory contains two folders:
 1. ngraph_ops - legacy opset operations needed for nGraph to CNNNetwork conversion.
 > **Note**: this operation are prohibited to use inside new plugins until they are not moved to separate directory with allowed operations.
 2. transformations - includes all transformations, utils, runtime info attributes and pass managers.
 > **Note**: do not use transformation that belongs to `ngraph::pass::ConvertOpSet1ToLegacy` transformations until they are not moved to separate directory with allowed transformations.
 
+TODO: show layer approach
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply docs/IE_PLUGIN_DG/NewTransformation.md:20]: Yes, I'll update them

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp (hunk 17 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.cpp:108]: If you want to to merge two GraphRewrite passes means that you can take MatcherPasses from them and combine manually.
pass::Manager responsible only for execution not combining. I think that in this case GraphRewrite `add_matcher` should support another GrpahRewrite instance as input and take matcher passes from it. But I still think that it's no good idea to merge two GraphRewrite passes

File: ngraph/src/ngraph/pass/pass.hpp (hunk 21 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/pass.hpp:42]: I hope that we can remove it in this release. I know that some plugins has NodePass transformations. So it will be removed in a separate request.

File: ngraph/src/ngraph/pass/pass.hpp (hunk 22 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/pass.hpp:100]: I'll move it to cpp

File: docs/template_plugin/src/template_pattern_transformation.hpp (hunk 19 above)
Comment (GlebKazantaev) [reply docs/template_plugin/src/template_pattern_transformation.hpp:23]: In nGraph we use Matcher class to store pattern and perform matching on node. So that's why I chose MatcherPass as a name for this class. May be it's better to have smth like PatternPass? @gladilov-gleb what do you think?

**Comment by gladilov-gleb** [comment]
@GlebKazantaev PR with 150+ changes and 60+ commits with messages without meaning (ex. `Codestyle` - what should I understand from it?) - makes review difficult and so decrease quality. Could you please prepare PR for a review first?

If it's okay with you, I think changes could be separated on introduction of updates in transformations API and applying those updates in specific transformations (it does not mean there should be just 2 commits!). Since the first part is the most interesting and should be reviewed with more attention, extracting it in a separate commit will help reviewers (at least me) concentrate on the main part. Please, explain more details in commit messages if it's possible. There is no restriction on message length and you could write rich text there about what and why has been done.

What I remember from discussion:

* @AlexKoff88 shared an opinion `MatcherPass` is not the best name (I don't have replacement right now)
* During discussion you had to spend some time explaining what returned `true` and `false` from API mean - it seems to me we could use self-descripting enum entries there in order to clarify intention. What do you think?

**Comment by AlexKoff88** [comment]
Possible options for `MatcherPass`:
- PattermMatcher
- TransformationMathcer
- MatchingTransformation

**Comment by GlebKazantaev** [comment]
> If it's okay with you, I think changes could be separated on introduction of updates in transformations API and applying those updates in specific transformations (it does not mean there should be just 2 commits and there could be more than 2 global parts!). Since the first part is the most interesting and should be reviewed with more attention, extracting it in a separate commit will help reviewers (at least me) concentrate on the main part. Please, explain more details in commit messages if it's possible. There is no restriction on message length and you could write rich text there about what and why has been done.

Together with @gladilov-gleb we tried to rewrite history and make a sequence of commits with meaningful description but due to this PR was created along ago and includes merges from master we decided to keep it as is.

> 
> What I remember from discussion:
> 
> * @AlexKoff88 shared an opinion `MatcherPass` is not the best name (I don't have replacement right now)
> * During discussion you had to spend some time explaining what returned `true` and `false` from API mean - it seems to me we could use self-descripting enum entries there in order to clarify intention. What do you think?

I was thinking about enums but it requires all passes to be updated. I will add it in a separate PR.

> * If now all transformations are registered in separate files - is there a way to understand which transformations and in which order will be applied (before I could use table files for that purpose)?
I just moved them to cpp files where you can see registration order as it was in tbl files.

**Comment by GlebKazantaev** [comment]
> Possible options for `MatcherPass`:
> 
> * PattermMatcher
> * TransformationMathcer
> * MatchingTransformation

In nGraph Matcher class is used for pattern matching on node, so for me `PatternMatcher` looks not natural. May be its better to call it like `PatternPass`

File: ngraph/src/ngraph/pass/pass.hpp (hunk 22 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/pass.hpp:100]: Done.

File: ngraph/src/ngraph/pass/manager.hpp (hunk 23 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/pass/manager.hpp:83]: I agree but in the same way nGraph uses lower case with underscore naming conventions for lambda types.

File: inference-engine/tests/unit/inference_engine/ngraph_graph_rewrite_test.cpp (hunk 24 above)
Comment (GlebKazantaev) [reply inference-engine/tests/unit/inference_engine/ngraph_graph_rewrite_test.cpp:1]: Moved to ngraph

File: ngraph/src/ngraph/pass/manager.hpp (hunk 23 above)
Comment (ilya-lavrenov) [reply ngraph/src/ngraph/pass/manager.hpp:83]: ok, good to know :)

**Comment by ilya-lavrenov** [comment]
Looks like the latest commits are not from Intel email :)

**Comment by GlebKazantaev** [comment]
> Looks like the latest commits are not from Intel email :)

OMG, you are right! Thanks!

File: docs/examples/example_ngraph_utils.cpp [hunk 26]
```diff
@@ -60,7 +60,7 @@ std::shared_ptr<ngraph::Function> create_advanced_function() {
 }
 // ! [ngraph_utils:advanced_function]
 
-void pattern_matcher_examples() {
+void pattern_matcher_examples(std::shared_ptr<Node> node) {
```
Comment (ilyachur) [inline docs/examples/example_ngraph_utils.cpp:63]: Why node is not constant? I didn't see any code which can change this pointer.

Also should we add reference to shared_ptr<Node>?

File: docs/template_plugin/src/template_function_transformation.cpp [hunk 27]
```diff
@@ -10,7 +10,7 @@ using namespace ngraph;
 
 // ! [function_pass:template_transformation_cpp]
 // template_function_transformation.cpp
-bool MyFunctionTransformation::run_on_function(std::shared_ptr<ngraph::Function> f) {
+bool pass::MyFunctionTransformation::run_on_function(std::shared_ptr<ngraph::Function> f) {
```
Comment (ilyachur) [inline docs/template_plugin/src/template_function_transformation.cpp:13]: std::shared_ptr&

File: docs/template_plugin/src/template_pattern_transformation.cpp [hunk 28]
```diff
@@ -43,9 +45,94 @@ void ngraph::pass::MyPatternBasedTransformation::transform() {
...
+    };
+
+    // Register pattern with Relu operation as a pattern root node
+    auto m = std::make_shared<ngraph::pattern::Matcher>(m_relu2, "ReluReluFusion");
+    // Register Matcher
+    this->register_matcher(m, callback);
+}
+// ! [matcher_pass:relu_fusion]
+
+void run_matcher_on_node(std::shared_ptr<ngraph::Node> node) {
```
Comment (ilyachur) [inline docs/template_plugin/src/template_pattern_transformation.cpp:88]: reference to shared_ptr

File: ngraph/test/pattern.cpp [hunk 29]
```diff
@@ -576,7 +577,7 @@ TEST(pattern, recurrent_pattern)
     std::set<std::shared_ptr<pattern::op::Label>> empty_correlated_matches;
     RecurrentMatcher rm(padd, rpattern, empty_correlated_matches);
     ASSERT_TRUE(rm.match(add3));
-    ASSERT_EQ(rm.get_number_of_bound_labels(), 1);
+    ASSERT_EQ(rm.get_number_of_bound_labels(), 3);
```
Comment (ilyachur) [inline ngraph/test/pattern.cpp:580]: why did you change the number of bound labels? How does it work?

Comment (GlebKazantaev) [reply ngraph/test/pattern.cpp:580]: Now we start capture local nodes. So that's an expected behavior.

File: docs/IE_PLUGIN_DG/NewTransformation.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply docs/IE_PLUGIN_DG/NewTransformation.md:20]: I removed TODOs but I'll finalyze this document in other PR.

File: docs/IE_PLUGIN_DG/NewTransformation.md (hunk 16 above)
Comment (GlebKazantaev) [reply docs/IE_PLUGIN_DG/NewTransformation.md:403]: done

File: docs/IE_PLUGIN_DG/NewTransformation.md (hunk 15 above)
Comment (GlebKazantaev) [reply docs/IE_PLUGIN_DG/NewTransformation.md:186]: I removed TODOs but I'll finalyze this document in other PR.

File: docs/examples/example_ngraph_utils.cpp (hunk 26 above)
Comment (GlebKazantaev) [reply docs/examples/example_ngraph_utils.cpp:63]: This function is used only to gather code in a single place and will be used only for docs generation.

File: docs/template_plugin/src/template_function_transformation.cpp (hunk 27 above)
Comment (GlebKazantaev) [reply docs/template_plugin/src/template_function_transformation.cpp:13]: This is the original function signature that I do not change.

File: docs/template_plugin/src/template_pattern_transformation.cpp (hunk 28 above)
Comment (GlebKazantaev) [reply docs/template_plugin/src/template_pattern_transformation.cpp:88]: Will be used only in docs and this function declaration won't be included.

File: inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_strided_slice_to_crop.cpp (hunk 20 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/convert_opset1_to_legacy/convert_strided_slice_to_crop.cpp:25]: The same comment as for MatcherPass

File: docs/IE_PLUGIN_DG/NewTransformation.md (hunk 18 above)
Comment (GlebKazantaev) [reply docs/IE_PLUGIN_DG/NewTransformation.md:418]: removed.
